src/transport_picoquic.cpp

```diff
@@ -0,0 +1,828 @@
+
+  picoquic_cnx_t* cnx = picoquic_create_cnx(quic_ctx,
+                                            picoquic_null_connection_id,
+                                            picoquic_null_connection_id,
+                                            (struct sockaddr*) & server_address,
+                                            current_time,
+                                            0, sni, config.alpn, 1);
+
+  if (cnx == NULL) {
+    logger.log(LogLevel::error, "Could not create picoquic connection client context");
+    return 0;
+  }
+
+  // Using default TP
+  picoquic_set_transport_parameters(cnx, &local_tp_options);
+
+  (void)createStreamContext(cnx, 0);
+
+  return (uint64_t)cnx;
+}
+
+void PicoQuicTransport::client(const TransportContextId tcid)
+{
+  int ret;
+  std::ostringstream log_msg;
+
+  picoquic_cnx_t* cnx = active_streams[tcid][0].cnx;
+
+  log_msg << "Thread client packet loop for client context_id: " << tcid;
+  logger.log(LogLevel::info, log_msg.str());
+
+  if (cnx == NULL) {
+    logger.log(LogLevel::error, "Could not create picoquic connection client context");
+  }
+  else {
+    picoquic_enable_keep_alive(cnx, 3000000);
+    picoquic_set_callback(cnx, pq_event_cb, this);
+
+    ret = picoquic_start_client_cnx(cnx);
+    if (ret < 0) {
+      logger.log(LogLevel::error, "Could not activate connection");
+      return;
+    }
+
+    (void)picoquic_mark_datagram_ready(cnx, true);
+
+    ret = picoquic_packet_loop(quic_ctx, 0, AF_INET, 0,
+                               2000000, 0,
+                               pq_loop_cb, this);
+
+    log_msg.str("");
+    log_msg << "picoquic ended with " << ret;
+    logger.log(LogLevel::info, log_msg.str());
+  }
+
+  if (quic_ctx != NULL) {
+    picoquic_free(quic_ctx);
+    quic_ctx = NULL;
+  }
+
+  shutdown();
+}
+
+void
+PicoQuicTransport::closeStream(const TransportContextId& context_id,
+                               const StreamId stream_id)
+{
+  deleteStreamContext(context_id, stream_id);
+}
+
+void
+PicoQuicTransport::close(
+        [[maybe_unused]] const TransportContextId& context_id)
+{
+}
+
+void PicoQuicTransport::checkDataOut()
+{
+  for (auto& c_pair: active_streams) {
+    for (auto &s_pair : active_streams[c_pair.first]) {
+
+//      std::ostringstream log_msg;
+//      log_msg << "pending data check, ctx: " << c_pair.first
+//              << " stream: " << s_pair.second.stream_id << " size: " << s_pair.second.out_data.size();
+//      logger.log(LogLevel::info, log_msg.str());
+      if (s_pair.second.out_data.size()) {
+          if (s_pair.first == 0) {
+            sendOutData(&s_pair.second, NULL, 1300);
+          } else {
+            (void)picoquic_mark_active_stream(s_pair.second.cnx,
+                                              s_pair.first,
+                                              1,
+                                              (StreamContext*)&s_pair.second);
+          }
+      }
+    }
+  }
```

Comment:
A lot of arrow code, and not using some of the nice stuff about about C++17/20 (i.e. structured-bindings). Makes it hard to read with all the `first`s and `second`s.

For example:
```cpp
for (auto& [_, active_stream] : active_streams) {
  for (auto& [ctx, data] : active_stream) {
    if (ctx == 0) {
      sendOutData(&data, NULL, 1300);
      continue;
    }
    // Awkward to format in GH editor so just imagine the params on newlines same as your code
    (void)picoquic_mark_active_stream(data.cnx, ctx, 1, reinterpret_cast<StreamContext*>(&data));
  }
}
```
Given it's a double for loop it is understandably hard to reduce the amount of tabbing, so arrow code is pretty much the only way.

Reply:
Just realized from reading other code further down I should've called `data` `stream` instead.